hpx/parallel/segmented_algorithms/transform.hpp

```diff
@@ -0,0 +1,876 @@
+                > is_seq;
+            typedef util::detail::algorithm_result<ExPolicy,
+                hpx::util::tagged_pair<tag::in(SegIter), tag::out(OutIter)>
+                > result;
+            if (first == last)
+            {
+                return result::get(std::make_pair(std::move(last),
+                    std::move(dest)));
+            }
+
+            typedef hpx::traits::segmented_iterator_traits<SegIter>
+                iterator_traits1;
+            typedef hpx::traits::segmented_iterator_traits<OutIter>
+                iterator_traits2;
+
+            return hpx::util::make_tagged_pair<tag::in, tag::out>(
+                segmented_transform(transform<std::pair<
+                typename iterator_traits1::local_iterator,
+                typename iterator_traits2::local_iterator> >(),
+                std::forward<ExPolicy>(policy), first, last,
+                dest, std::forward<F>(f), proj, is_seq()));
+        }
+
+        // forward declare the non-segmented version of this algorithm
+        template <typename ExPolicy, typename InIter, typename OutIter, typename F,
+            typename Proj>
+        typename util::detail::algorithm_result<ExPolicy,
+            hpx::util::tagged_pair<tag::in(InIter), tag::out(OutIter)>
+        >::type
+        transform_(ExPolicy&& policy, InIter first, InIter last, OutIter dest,
+            F && f, Proj && proj, std::false_type);
+
+        // Binary transform
+
+        // sequential remote implementation
+        template <typename Algo, typename ExPolicy, typename InIter1,
+            typename InIter2, typename OutIter, typename F,
+            typename Proj1, typename Proj2>
+        static typename util::detail::algorithm_result<ExPolicy,
+            hpx::util::tuple<InIter1,InIter2,OutIter>
+        >::type
+        segmented_transform(Algo && algo, ExPolicy const& policy,
+            InIter1 first1, InIter1 last1, InIter2 first2, OutIter dest,
+            F && f, Proj1 && proj1, Proj2 && proj2, std::true_type)
+        {
+            typedef hpx::traits::segmented_iterator_traits<InIter1> traits1;
+            typedef hpx::traits::segmented_iterator_traits<InIter2> traits2;
+            typedef hpx::traits::segmented_iterator_traits<OutIter> traits3;
+            typedef typename traits1::segment_iterator segment_iterator1;
+            typedef typename traits1::local_iterator local_iterator_type1;
+            typedef typename traits2::segment_iterator segment_iterator2;
+            typedef typename traits2::local_iterator local_iterator_type2;
+            typedef typename traits3::segment_iterator segment_iterator3;
+            typedef typename traits3::local_iterator local_iterator_type3;
+            typedef util::detail::algorithm_result<ExPolicy,
+                hpx::util::tuple<InIter1,InIter2,OutIter>
+            > result;
+            auto last2 = first2;
+            detail::advance(last2,std::size_t(last1 - first1));
```

Comment:
Should I restrict the iterators to be random iterators using a check or is there a better more generic way to find the last (end) iterator given the first (begin) iterator?